File: hash.hs
import System.IO
import Data.List

data Pair   = N | E {key::Int, val::Int} deriving(Show, Eq)

--PROMPT ROUTINES----------------------------------------------------------------------------------------------

prompt :: String -> IO ()
prompt str      = (putStr str)      >>= (\x -> hFlush stdout)

promptLn :: String -> IO ()
promptLn str    = (putStrLn str)    >>= (\x -> hFlush stdout)

--HELPER ROUTINES----------------------------------------------------------------------------------------------

hashcode :: Int -> Int -> Int
hashcode key size
    | (size <= 1)       = 0
    | otherwise         = rem key size

hindex :: Int -> Int -> Int -> Int
hindex key trial size = rem ((hashcode key size) + (trial * (hashcode key s))) size where s = (div size 2) + 7

weight :: [Pair] -> Int
weight []       = 0
weight (p:ps)
    | p == N    = weight ps
    | otherwise = 1 + (weight ps)

trv :: [Pair] -> String
trv []                  = "!! Empty Array"
trv [N]                 = "[_:_]"
trv [(E key val)]       = "[" ++ (show key) ++ ":" ++ (show val) ++ "]"
trv (N:ps)              = "[_:_] " ++ trv ps
trv ((E key val):ps)    = "[" ++ (show key) ++ ":" ++ (show val) ++ "] " ++ trv ps

--CREATE ROUTINES----------------------------------------------------------------------------------------------

makeArray :: Int -> [Pair]
makeArray 0             = []
makeArray size          = N : (makeArray $ size - 1)

--ACTION ROUTINES----------------------------------------------------------------------------------------------

put :: [Pair] -> Int -> Int -> [Pair]
put ps k v
    | k < 0 = ps
    | (weight (add ps 0 k v)) <= (div (length ps) 2) = add ps 0 k v
    | otherwise = extend (add ps 0 k v) (makeArray (2 * (length ps)))

add :: [Pair] -> Int -> Int -> Int -> [Pair]
add ps t k v 
    | let h = hindex k t (length ps), (head $ drop h ps) == N = (take h ps) ++ [(E k v)] ++ (drop (h+1) ps)
    | let h = hindex k t (length ps), (key $ head $ drop h ps) == k = ps
    | t > (25 * (length ps)) = (makeArray (length ps))
    | otherwise = add ps (t+1) k v

extend :: [Pair] -> [Pair] -> [Pair]
extend [] ns    = ns
extend (p:ps) ns
    | p == N    = extend ps ns
    | otherwise = extend ps (put ns (key p) (val p))

pop :: [Pair] -> Int -> [Pair]
pop ps k
    | (weight (del ps 0 k)) > (div (length ps) 4) = del ps 0 k
    | length ps <= 8 = del ps 0 k
    | otherwise = shrink (del ps 0 k) (makeArray (div (length ps) 2))

del :: [Pair] -> Int -> Int -> [Pair]
del ps t k
    | let h = hindex k t (length ps), (head $ drop h ps) == N = ps
    | let h = hindex k t (length ps), (key $ head $ drop h ps) == k = (take h ps) ++ [N] ++ (drop (h+1) ps)
    | t > (25 * (length ps)) = (makeArray (length ps))
    | otherwise = del ps (t+1) k

shrink :: [Pair] -> [Pair] -> [Pair]
shrink [] ns    = ns
shrink (p:ps) ns
    | p == N    = shrink ps ns
    | otherwise = shrink ps (put ns (key p) (val p))

finder :: [Pair] -> Int -> String
finder ps k = search ps 0 k

search :: [Pair] -> Int -> Int -> String
search [] _ _ = "No such key"
search ps t k
    | let h = hindex k t (length ps), (head $ drop h ps) == N = "No such key"
    | let h = hindex k t (length ps), (key $ head $ drop h ps) == k = "Found the key " ++ (show $ k) ++ " and its value " ++ (show $ val $ head $ drop h ps)
    | otherwise = search ps (t+1) k

--MAIN ROUTINE-------------------------------------------------------------------------------------------------

hash :: [Pair] -> IO()
hash ps = do
    prompt ">> Enter a command: "
    cmd <- getLine
    case cmd of 
        "exit"      -> 
            return()
        "reset"     -> 
            (promptLn ">> Resetting the array...")
            >>= (\x -> (promptLn $ trv []))
            >>= (\x -> (hash []))
        "make"      -> do
            prompt ">> Enter a size of the array: "
            s <- getLine
            let size = read s::Int
            promptLn $ trv $ makeArray size
            hash $ makeArray size
        "new"       ->
            (promptLn ">> Creating an empty array with length 7")
            >>= (\x -> (promptLn $ trv $ makeArray 7))
            >>= (\x -> (hash $ makeArray 7))
        "put"       -> do
            if (length ps) < 1 then do 
                promptLn "!! Please create an empty array using <make> or <new> first"
                hash ps
            else do
                promptLn ">> Putting a key-val pair to the list"
                prompt "  $ Enter a key to put: "
                k <- getLine
                prompt "  $ Enter a corresponding value: "
                v <- getLine
                let key  = read k::Int
                let val  = read v::Int        
                promptLn $ trv $ put ps key val   
                hash     $ put ps key val
        "pop"       -> do
            if (length ps) < 1 then do 
                promptLn "!! Please create an empty array using <make> or <new> first"
                hash ps
            else do
                prompt "  $ Enter a key to remove: "
                k <- getLine
                let key = read k::Int
                promptLn $ trv $ pop ps key
                hash     $ pop ps key
        "trv"       ->  do
            promptLn $ trv ps
            hash ps
        "find"      -> do
            promptLn ">> Find the key and get its value"
            prompt "  $ Enter a key to find: "
            k <- getLine
            let key  = read k::Int
            promptLn $ "  # " ++ (finder ps key)
            hash ps
        "--help"    -> do
            promptLn "+---------------------------------------------------------------+"
            promptLn "| $ exit  - exit the program                                    |"                                   
            promptLn "| $ reset - reset the list                                      |"
            promptLn "| $ make  - make an empty list                                  |"
            promptLn "| $ new   - make an empty list with length 7                    |"
            promptLn "| $ put   - put a key-value pair to the list                    |"
            promptLn "| $ pop   - remove a key-value pair from the list               |"
            promptLn "| $ find  - find a key-value pair from the list                 |"
            promptLn "| $ trv   - traverse the list                                   |"
            promptLn "+---------------------------------------------------------------+"
            hash ps
        _           -> do
            promptLn $ "!! <" ++ cmd ++ "> is not a valid command"
            promptLn $ "!! use --help to see the list of commands"
            hash ps

main :: IO ()
main = do
    promptLn "-----------------------------------------------------------------"
    promptLn "Hash with Haskell"
    promptLn "## use --help to see the list of commands"
    hash []
    promptLn "-----------------------------------------------------------------"

File: Hash/hash.py
VERBOSE = 0

def hashcode(k, s):
    return k%s

def hIndex(k, t, s):
    return hashcode(hashcode(k,s) + t*hashcode(k,int((s+1)*7)), s)

class Pair:
    def __init__(self, k, v):
        self.key = k
        self.val = v

    def setKey(self, k):
        self.key = k
        return

    def setVal(self, v):
        self.val = v
        return
    
    def getKey(self):
        return self.key
    
    def getVal(self):
        return self.val

    def show(self):
        print("[%d:%d]" % (self.key, self.val))

class Hash:
    def __init__(self, s):
        self.size = s
        self.weight = 0
        self.arr = []
        for i in range(s):
            self.arr.append(None)
       
    def put(self, k, v):
        t = 0
        while (t < self.size*20):
            h = hIndex(k, t, self.size)
            if (self.arr[h] == None):
                self.arr[h] = Pair(k, v)
                self.weight += 1
                if VERBOSE == 1:
                    print("weight is %d" % (self.weight))
                break
            else:
                if (self.arr[h].getKey() == k):
                    print(">> no duplicate keys allowed!")
                    return
                t = t + 1
        if self.weight > int(self.size/2):
            print(">> key (%d) added -> extending the list..." % (k))
            self.extend()
        return
    
    def extend(self):
        newarr = []
        s = self.size
        for i in range(s):
            newarr.append(self.arr[i])
        self.arr = []
        for i in range(s*2):
            self.arr.append(None)
        self.size = 2*s
        self.weight = 0
        for i in range(s):
            if newarr[i] != None:
                self.put(newarr[i].getKey(), newarr[i].getVal())
        return

    def pop(self, k):
        t = 0
        while (t < self.size*20):
            h = hIndex(k, t, self.size)
            if (self.arr[h] != None):
                if (self.arr[h].key == k):
                    self.arr[h] = None
                    self.weight -= 1
                    break
                else:
                    t = t + 1
            else:
                print("!! no such key (%d) found" & (k))
                return
        if int(self.size/2) >= 5:
            if self.weight <= int(self.size/4):
                print(">> diminishing the list...")
                self.diminish()
        return
    
    def diminish(self):
        newarr = []
        s = self.size
        for i in range(s):
            newarr.append(self.arr[i])
        self.arr = []
        for i in range(int(s/2)):
            self.arr.append(None)
        self.size = int(s/2)
        self.weight = 0
        for i in range(s):
            if newarr[i] != None:
                self.put(newarr[i].getKey(), newarr[i].getVal())
        return

    def traverse(self):
        for h in range(self.size):
            if self.arr[h] == None:
                print("[_:_]", end = " ")
            else:
                print("[%d:%d]" % (self.arr[h].key, self.arr[h].val), end = " ")
        print("")
        return
    

h1 = Hash(11)
h1.traverse()
for i in range(9):
    h1.put(3*i,3*i)
h1.traverse()
for i in range(0,8,2):
    h1.pop(3*i)
h1.traverse()
h1.pop(9)
h1.pop(21)
h1.pop(15)
h1.traverse()
